src/core/server/server.ts

```diff
@@ -113,27 +135,78 @@ export class Server {
     this.executionContext = new ExecutionContextService(core);
+    this.prebootService = new PrebootService(core);
 
     this.savedObjectsStartPromise = new Promise((resolve) => {
       this.resolveSavedObjectsStartPromise = resolve;
     });
   }
 
-  public async setup() {
-    this.log.debug('setting up server');
-    const setupTransaction = apm.startTransaction('server_setup', 'kibana_platform');
+  public async preboot() {
+    this.log.debug('prebooting server');
+    const prebootTransaction = apm.startTransaction('server_preboot', 'kibana_platform');
 
-    const environmentSetup = await this.environment.setup();
+    this.environmentPreboot = await this.environment.preboot();
 
     // Discover any plugins before continuing. This allows other systems to utilize the plugin dependency graph.
-    const { pluginTree, pluginPaths, uiPlugins } = await this.plugins.discover({
-      environment: environmentSetup,
+    this.discoveredPlugins = await this.plugins.discover({ environment: this.environmentPreboot });
+
+    // Immediately terminate in case of invalid configuration. This needs to be done after plugin discovery. We also
+    // silent deprecation warnings until `setup` stage where we'll validate config once again.
+    await ensureValidConfiguration(this.configService, { logDeprecations: false });
+
+    const { uiPlugins, pluginTree, pluginPaths } = this.discoveredPlugins.preboot;
+    const contextServiceSetup = this.context.setup({
+      pluginDependencies: new Map([...pluginTree.asOpaqueIds]),
+    });
+    const httpPreboot = await this.http.preboot({ context: contextServiceSetup });
+
+    // setup i18n prior to any other service, to have translations ready
+    await this.i18n.preboot({ http: httpPreboot, pluginPaths });
+
+    this.capabilities.preboot({ http: httpPreboot });
+    const elasticsearchServicePreboot = await this.elasticsearch.preboot();
+    const uiSettingsPreboot = await this.uiSettings.preboot();
+    await this.status.preboot({ http: httpPreboot });
+
+    const renderingPreboot = await this.rendering.preboot({ http: httpPreboot, uiPlugins });
+    const httpResourcesPreboot = this.httpResources.preboot({
+      http: httpPreboot,
+      rendering: renderingPreboot,
     });
 
-    // Immediately terminate in case of invalid configuration
-    // This needs to be done after plugin discovery
+    this.loggingSetup = this.logging.setup({ loggingSystem: this.loggingSystem });
+
+    const corePreboot: InternalCorePreboot = {
+      context: contextServiceSetup,
+      elasticsearch: elasticsearchServicePreboot,
+      http: httpPreboot,
+      uiSettings: uiSettingsPreboot,
+      httpResources: httpResourcesPreboot,
+      logging: this.loggingSetup,
+      preboot: this.prebootService.preboot(),
+    };
+
+    await this.plugins.preboot(corePreboot);
+
+    httpPreboot.registerRouteHandlerContext(coreId, 'core', (() => {
+      return new PrebootCoreRouteHandlerContext(corePreboot);
+    }) as any);
```

Comment:
You mean `any` or the `httpPreboot.registerRouteHandlerContext` call in general? If the latter, then we need it for the `HttpResourceService` to [render the app](https://github.com/elastic/kibana/blob/4a541883557bcee83baf09b2c0ddab702f780e45/src/core/server/http_resources/http_resources_service.ts#L97). That's actually the only reason in having this hacky `UiSettingsClient` at all, I tried to refactor this to avoid the default client and this context altogether but it led to much more changes in our rendering pipeline that I was comfortable to make in this PR.

Reply:
@mshustov @pgayvallet the only other alternative to this messy request handler and the hacky UiSettingsClient I could come up with is this https://github.com/elastic/kibana/commit/7ad8b6958f150283575716dd8fbf349a5e1de46e (draft):

Instead of maintaining a semi-fake "default" UiSettingsClient, we can construct `coreDefaultUiSettings` at `preboot` and rely on it during rendering when scoped UiSettingsClient isn't available (at `preboot` basically). As a consequence scoped UI settings client becomes optional in a bunch of places. I'm hesitating to omit `coreDefaultUiSettings` entirely since we have a bunch of default core settings that may be useful for `preboot` applications (date format, notification settings to name a few).

I like none of these options, but this one is probably easier to understand when reading the code. What do you think? Do you have any better idea how we can hop through this ubiquitous `uiSettingsClient` dependency?

Reply:
> Do you have any better idea how we can hop through this ubiquitous uiSettingsClient dependency?

Another alternative that I'd suggest might be to implement `IUiSettingsClient` interface, but with `noop` method for the functionality we don't support (`write` operations, for example).

Reply:
> Another alternative that I'd suggest might be to implement IUiSettingsClient interface, but with noop method for the functionality we don't support (write operations, for example).

Sure, let me try that then, we'll most likely just have `getRegistered` and maybe `get/getAll` that will use `getRegistered` under the hood.

Reply:
Done in https://github.com/elastic/kibana/pull/103636/commits/c3b8ede0ec9247dfcd8a90003451314814fae1bb